exercises/src/test/scala/datastructures/ListSpec.scala

```diff
@@ -0,0 +1,417 @@
+import org.specs2.mutable.Specification
+import org.specs2.ScalaCheck
+
+import fpinscala.datastructures._
+
+class ListSpec extends Specification {
+  //========================================================================================
+  //#3.1
+  "pattern matching example x" should {
+    "result in 3" in {
+      List.x mustEqual 3
+    }
+  }
+
+  //========================================================================================
+  //#3.2
+  "tail" should {
+    "return Nil for an empty List" in {
+      List.tail(List()) mustEqual Nil
+    }
+
+    "return Nil for single-element List" in {
+      List.tail(List(1)) mustEqual Nil
+    }
+
+    "return tail of multiple-element List" in {
+      List.tail(List(1,2)) mustEqual List(2)
+      List.tail(List('a','b')) mustEqual List('b')
+    }
+  }
+
+  //========================================================================================
+  //#3.3
+  "setHead" should {
+    "update nothing for an empty List" in {
+      List.setHead(List(), 0) mustEqual Nil
+    }
+
+    "update the head for single-element List" in {
+      List.setHead(List(1), 0) mustEqual List(0)
+      List.setHead(List('a'), 'z') mustEqual List('z')
+    }
+
+    "update head for multiple-element List" in {
+      List.setHead(List(1,2), 0) mustEqual List(0,2)
+      List.setHead(List('a','b'), 'z') mustEqual List('z','b')
+    }
+  }
+
+  //========================================================================================
+  //#3.4
+  "drop" should {
+    "result in Nil for empty List" in {
+      List.drop(List(), 2) mustEqual Nil
```

Comment:
IMO, trying to drop more elements than the list contains should throw an error. It's like trying to take the head or tail of an empty list.